### ISSUE TITLE
Prebid doesn't wait with requests for close the GDPR window

### ISSUE THREAD

**Issue by modessef1** [prebid/Prebid.js#5694]
## Type of issue
question

## Description
We have a problem with the prebid not waiting for the GDPR window to close or until the set timeout is reached. It immediately sends the request and starts GAM. Could it be because when we display the GDPR window we have gdprApplies: false and only after closing the window gdprApplies changes to true?
If it matters, how to control it? We tried consentManagement.gdpr.defaultGdprScope: true but it didn't help. If it doesn't matter what should we check first?
We use the CMP written by us.

## Test page
https://parkiet.com

## Platform details
Prebid.js 4.3
TCF 2.0

**Comment by bretg** [comment]
PBJS calls the __tcfapi() function if it exists -- if it gets a response right away, it starts the auction. To slow it down, you set the `timeout` value.

You need to follow the CMP spec defined by the IAB. https://github.com/InteractiveAdvertisingBureau/iabtcf-es

Here's a complicated flowchart I did for how this works. @Fawke - can you review it please? I think something may have changed recently around Purpose1Treatment

![PBJS TCF2](https://user-images.githubusercontent.com/4412838/92161917-506d3880-edff-11ea-97cc-2548c67006af.png)

**Comment by modessef1** [comment]
we understand how it works. In prebid 3.x.x that worked for us, 4.x.x doesn't work anymore and we don't know why.

**Comment by bretg** [comment]
The difference between 3.x and 4.x is that gdpr enforcement for all purposes is on by default. Specifically, the "basicAds" (Purpose 2) is on. If you want auctions to happen even if the user has not yet provided consent even in the EEA, you would have to turn off the enforcement for "basicAds". 

{code}
pbjs.setConfig({
  consentManagement: {
    gdpr: {
      cmpApi: 'iab',   // activates the enforcement module
      defaultGdprScope: true,
      timeout: 10000; //default value
      rules: [{        // these are the default values
        purpose: "storage",
        enforcePurpose: true,
        enforceVendor: true
      },{
        purpose: "basicAds",
        enforcePurpose: true,
        enforceVendor: true
      },{
        purpose: "measurement",
        enforcePurpose: true,
        enforceVendor: true
      }]
    }
  }
});
{code}


> prebid not waiting for the GDPR window to close or until the set timeout is reached

Prebid.js waits for the consentManagement.gdpr.timeout. It certainly doesn't wait for the GDPR window. By "set timeout", do you mean the requestBids auction timeout? That's not part of this either.

You page does set consentManagement.gdpr.timeout=10000, which means that Prebid.js should wait 10 seconds for the __tcfapi call to respond. Are you saying that's not happening?  @Fawke or @robertrmartinez  - any changes between 3.x and 4.x with regards to how waiting for __tcfapi works?

**Comment by modessef1** [comment]
ok, that explains that the way the module works has changed but this has not been changed in the documentation.
The module is now not waiting for approval, although it is in the documentation:
`In the case of a new user, CMPs will generally respond only after there is consent information available (i.e., the user has made their consent choices). Making these selections can take some time for the average user, so the module provides timeout settings.`
just to load the popup.
So we have to write it ourselves or have enforcement turned on

**Comment by bretg** [comment]
> The module is now not waiting for approval

It should be waiting up to the consentManagement.gdpr.timeout value.

When you hit https://parkiet.com?pbjs_debug=true, what messages do you see that indicate it's not waiting? I see this message immediately even though I delete cookies and local storage and don't touch the CMP popup window:

```
INFO: Received a response from CMP 
```

If you see a "CMP workflow exceeded timeout threshold" message immediately, that would indicate the timeout isn't working.

**Comment by robertrmartinez** [comment]
@modessef1 @bretg 

The "issue" here is that your tcfApi "cmpUiIsShown" response is returning gdprApplies === false:

![image](https://user-images.githubusercontent.com/13972794/92262377-ddf86900-ee8f-11ea-9b74-a559618fe1aa.png)


So following the logic in the code:

```js
  function v2CmpResponseCallback(tcfData, success) {
    utils.logInfo('Received a response from CMP', tcfData);
    if (success) {
      if (tcfData.gdprApplies === false) {
        cmpSuccess(tcfData, hookConfig);
      } else if (tcfData.eventStatus === 'tcloaded' || tcfData.eventStatus === 'useractioncomplete') {
        cmpSuccess(tcfData, hookConfig);
      } else if (tcfData.eventStatus === 'cmpuishown' && tcfData.tcString && tcfData.purposeOneTreatment === true) {
        cmpSuccess(tcfData, hookConfig);
      }
    } else {
      cmpError('CMP unable to register callback function.  Please check CMP setup.', hookConfig);
    }
  }
```

We are under the first IF statement.

If the CMP tells us GDPR does not apply, then we will not wait around for more responses, we will continue with the auction setting gdpr = false.

Setting `defaultGdprScope` to `true` does not do anything for you because this param only is used if we were unable to get any response from the CMP.

If a cmp tells us gdprApplies is false, we use that.

Of course I do not know the inner workings of your self implemented CMP,

But if you want it to wait, try setting gdprApplies to `true` in your `cmpuiisshown` callback.

I do not know if it is set dynamically based on a geo lookup or anything.

**Comment by robertrmartinez** [comment]
Note: The above snippet of code has been altered in latest master.

(Although it does not change how it would work on the example page you provided)

```js
  function v2CmpResponseCallback(tcfData, success) {
    utils.logInfo('Received a response from CMP', tcfData);
    if (success) {
      if (tcfData.gdprApplies === false || tcfData.eventStatus === 'tcloaded' || tcfData.eventStatus === 'useractioncomplete') {
        cmpSuccess(tcfData, hookConfig);
      }
    } else {
      cmpError('CMP unable to register callback function.  Please check CMP setup.', hookConfig);
    }
  }
```

You still would fall under the first check `tcfData.gdprApplies === false`

And thus, the consent module will exit and signal gdprApplies is false.

**Comment by bretg** [comment]
I tried VPNing to the EEA, removing all cookies and local storage and still received gdpApplies:false from the CPM immediately

**Comment by Fawke** [comment]
> PBJS calls the __tcfapi() function if it exists -- if it gets a response right away, it starts the auction. To slow it down, you set the `timeout` value.
> 
> You need to follow the CMP spec defined by the IAB. https://github.com/InteractiveAdvertisingBureau/iabtcf-es
> 
> Here's a complicated flowchart I did for how this works. @Fawke - can you review it please? I think something may have changed recently around Purpose1Treatment
> 
> ![PBJS TCF2](https://user-images.githubusercontent.com/4412838/92161917-506d3880-edff-11ea-97cc-2548c67006af.png)

Hi Bret,

This flowchart looks good, it just need one change. We're no longer listening to `cmpuishown` event. So, the check, `tcString exists and Purpose1 treatmean = 1` needs to go.

**Comment by smenzer** [comment]
FWIW...I just checked the test page (https://parkiet.com?pbjs_debug=true) from France and the CMP response shows `gdprApplies: true` in all responses:

![image](https://user-images.githubusercontent.com/624392/92401384-d524c300-f12d-11ea-938f-b413dcb843c5.png)

**Comment by modessef1** [comment]
This morning we released a fix to our CMP. Previously It always returned gdprApplies: false on startup. It's okay now.
